File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    queue = deque()
    queue.append([0,0])
    dir = [[0,1], [0,-1], [1,0], [-1,0]]
    n = len(maps)-1
    m = len(maps[0])-1
    while queue:
        cur = queue.popleft()
        for d in dir:
            dx, dy = cur[0] + d[0], cur[1] + d[1]
            if dx>n or dy>m or dx<0 or dy<0 or maps[dx][dy]!=1:
                continue
            maps[dx][dy] += maps[cur[0]][cur[1]]
            queue.append([dx, dy])
    print(maps)
    if maps[n][m] == 1:
        return -1
    return maps[n][m]

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        int[] cur = {0,0};
        int n = maps.length;
        int m = maps[0].length;

        int[][] dirs = {{0,1},{0,-1},{1,0},{-1,0}};
        Queue<int[]> queue = new ArrayDeque<>();
        int[][] visited = new int[n][m];

        for(int i=0; i<n; i++) {
            for (int j=0; j<m; j++) {
                visited[i][j] = -1;
            }
        }

        queue.offer(cur);
        visited[cur[0]][cur[1]] = 1;

        while(!queue.isEmpty()) {
            cur = queue.poll();
            int x = cur[0];
            int y = cur[1];

            for (int[] dir : dirs) {
                int dx = x + dir[0];
                int dy = y + dir[1];

                if (dx>=0 && dx<n && dy>=0 && dy<m && visited[dx][dy]==-1 && maps[dx][dy] == 1) {
                    queue.offer(new int[] {dx, dy});
                    visited[dx][dy] = visited[x][y]+1;
                }
            }
        }
        int answer = visited[n-1][m-1];
        return answer;
    }
}